### PR TITLE
Offscreen RPC

### PR DESCRIPTION
**Pull request by turbocrime**
# offscreen rpc

did some exploratory work today attempting to host rpc in the offscreen document.

## why

this is intended to make progress towards #341

targeting native dom will simplify our architecture, make our rpc implementations more portable, and also force us to abstract our context interfaces, which will make our implementations less prax-specific.

simplification alone is enough of a motivator. what actually took most of my time today was pairing with @jessepinho on worker errors. the current architecture is nearly impossible to debug: build logic is split across five documents and several simultaneous wasm runtimes, through more layers of ser/de than i want to count ;_;

## current situation

welcome to hell!! welcome to hell!!

```mermaid
graph TB

Minifront

subgraph Prax



        style storage color:red,stroke:red
        subgraph storage
            pk((Password Key))
            subgraph localExt
            wallets((wallets))
            idbConstants
            grpcEndpoint
            end
        end

        page[Extension Pages]

        subgraph service_worker.ts
            init>services.initialize]
            listeners{{Listeners}}
            SessionManager

            subgraph Services
            style Services fill:green
                queriers
                subgraph WalletServices

                    subgraph ViewServer
                        fvk((Full Viewing Key))
                        esp((Encrypted Seed Phrase))
                        wview[/WASM ViewServer\]
                    end

                    idb{{IndexedDbInterface}}
                    BlockProcessor 
                end
            end

            subgraph router [ConnectRouter]
            style router fill:purple
                subgraph impls
                view{View RPC}
                custody{Custody RPC}
                proxy{Proxy RPC}

                wplan[/WASM Planner\]
                wbuild[/WASM Builder\]
                offscreenClient{{Offscreen Client<br>not context, not an impl, but here anyway :C}}
                end
            end
        end



    subgraph offscreen
        spawnWorker
        w1[\wasm/]
        w2[\wasm/]
        w3[\wasm/]
    end

    IndexedDb[(IndexedDb)]
end

pd{remote pd}


idb === IndexedDb
classDef idbClass stroke:cyan,color:cyan,fill:blue
class idb,IndexedDb idbClass
BlockProcessor === idb 
view ----->|ctx| idb 
wbuild & wplan & wview -.-> IndexedDb
linkStyle 0,1,2,3,4,5 stroke:cyan;

view ==o|view.wallet-id| wallets
custody ==o|custody.authorize| pk 
custody ==o|custody.authorize| wallets
linkStyle 6,7,8 stroke:red,color:red

page <-.-> storage

init ===o queriers & idb & WalletServices
init -.-o grpcEndpoint & wallets & idbConstants 
linkStyle 10,11,12,13,14,15 stroke:lime,color:lime

Minifront <-.->|content scripts| listeners <--> SessionManager <-->|adapter| router
page <-.|pageClient|.-> router
linkStyle 16,17,18,19 stroke:magenta


fvk --o view & custody
esp --o custody
linkStyle 20,21,22 stroke:lime

view ==> offscreenClient & wplan & wbuild

offscreenClient <-.-> spawnWorker <==> w1 & w2 & w3  

BlockProcessor === wview
BlockProcessor ==== queriers 
fvk ==o wview & BlockProcessor


view =====>|ctx| custody -."ctx<br>popup".-> page
view ----->|ctx| queriers


queriers --> pd & pd & pd

wview --- queriers
proxy --> pd


classDef localExtStorage color:black,fill:yellow;
class wallets,idbConstants,grpcEndpoint localExtStorage
classDef sessionExtStorage color:black,fill:orange;
class pk sessionExtStorage

classDef offscreenBoundary fill:purple
class offscreen offscreenBoundary

classDef serviceInit stroke:lime
class init serviceInit

classDef wasm fill:teal,color:orange;
class wview,wbuild,wplan,w1,w2,w3 wasm

classDef transport color:magenta
class view,custody,proxy transport

view ~~~ custody
init ~~~ storage
```

tl;dr the colored lines crossing the purple are bad.

green, 'Services' key access, should just happen on init.

cyan, idb access, should be fine if we can access idb in the offscreen document. if we can't, ts can easily run queries through a chrome runtime connection, but wasm idb access may be problematic.

red, storage access, may be mostly replaceable in some cases by the green key access.

### i did not do any of that work yet

but this feels like good progress towards understanding what's necessary

### but i did achieve some reorganization

the simplified event/message flow for builds should turn out very nice - workers will launch straight from `optimisticBuild`, instead of through multiple layers of indirection

https://github.com/penumbra-zone/web/blob/040d18e6337d4ace853098d5ff966588db82392b/packages/router/src/grpc/view-protocol-server/util/build-tx.ts#L35-L46

and from the same file:

https://github.com/penumbra-zone/web/blob/040d18e6337d4ace853098d5ff966588db82392b/packages/router/src/grpc/view-protocol-server/util/build-tx.ts#L89-L114

## what are the obstacles to offscreen rpc?

1. offscreen may have limitations on things like idb access. unfortunately, documentation is limited. i'll investigate next week
2. if idb limitations exist, it may be difficult to work around them inside wasm
3. launching things like the popup might require one more step of message-passing

## what is not an obstacle to offscreen rpc?

1. offscreen cannot access extension storage. but the impls access it in very few places, and this needs to be changed anyway.
2. if idb limitations exist, we have already developed the tools to efficiently stream queries across the chrome runtime.
3. rpc messaging should be exactly identical. the same tools we use to cross the chrome runtime now will be suitable for use in offscreen.

### CODE REVIEW

File: apps/extension/src/entry/penumbra-rpc.ts
Comment (turbocrime) [inline apps/extension/src/entry/penumbra-rpc.ts:1]: initial attempt at this included 'services' in the rpc offscreen. this probably isn't necessary but it was quick to try

**Comment by turbocrime** [comment]
#488 would obviate wasm-idb concerns

**Comment by turbocrime** [comment]
confirmed that offscreen has no problem accessing/writing to the same idb as the rest of the extension

**Comment by Valentine1898** [comment]
Is this PR supposed to solve issue #488?
